### PR TITLE
test: centralize aws sigv4 test fixtures

### LINKED ISSUE #20785: refactor(mitm-addon): centralize AWS SigV4 test fixtures
## Background

`crates/runner/mitm-addon/tests/test_firewall_aws_sigv4_auth.py` has grown to 1,388 lines and repeats the same AWS SigV4 request/auth scaffolding throughout the file. The repeated setup includes standard STS/IAM `real_flow(...)` construction, placeholder `Authorization` headers, presigned query strings, AWS auth config/auth response payloads, and cached `AwsSigV4Credentials` setup for fail-closed paths.

Related duplicated SigV4 fixture vocabulary also exists in neighboring tests:

- `crates/runner/mitm-addon/tests/test_aws_sigv4.py`
- `crates/runner/mitm-addon/tests/test_firewall_auth.py`
- `crates/runner/mitm-addon/tests/test_firewall_rewrite_forwarding.py`

The current duplication makes future changes to canonical SigV4 fixture values, placeholder credential shape, standard auth response shape, or cache setup noisy and easy to apply inconsistently. This is especially risky because SigV4 auth is security-sensitive: tests should make behavior differences obvious, not bury them in repeated fixture boilerplate.

## Updated Decision

This should be treated as a test-maintenance refactor, not a production behavior change. Production code should not be changed for this issue.

The better direction is not just to add a few local helpers to `test_firewall_aws_sigv4_auth.py`. The file is also carrying too much of the SigV4 behavior matrix in one place. The refactor should unify the fixture language and, if it improves readability, split the large test file by behavior area.

## Proposed Structure

Use two layers of test helpers instead of one overly broad helper module.

### 1. Pure SigV4 fixture helpers

Add `crates/runner/mitm-addon/tests/aws_sigv4_helpers.py` for pure SigV4 test vocabulary only:

- Placeholder access key ID and credential scope constants.
- Placeholder header `Authorization` builder.
- Placeholder presigned query path/URL builder.
- Standard STS/IAM request constants.
- Real test `AwsSigV4Credentials` values.

This module should not know about firewall auth endpoint responses, cache keys, `FirewallAllow`, or `auth.handle_firewall_request(...)`.

### 2. Firewall integration helpers

Add `crates/runner/mitm-addon/tests/firewall_aws_sigv4_helpers.py` for firewall-auth integration setup:

- Standard AWS auth config objects.
- Standard auth endpoint responses, with and without a session token.
- `vm_info` and `FirewallAllow` builders for AWS tests.
- Cached credential setup for fail-closed paths.
- A small wrapper that preserves the real `auth.handle_firewall_request(...)` integration path.

This keeps signer-level fixtures reusable without coupling all SigV4 tests to firewall-auth internals.

## Test File Organization

After the helper extraction, consider splitting `test_firewall_aws_sigv4_auth.py` if the resulting file is still hard to scan. A reasonable split would be:

- Header re-sign success tests.
- Query re-sign success tests.
- Header fail-closed tests.
- Query fail-closed tests.
- Trusted original URL / rewrite-adjacent edge cases.

The split is not required if helper extraction makes the file readable enough, but the implementation should not be constrained to the smallest possible diff.

## Parameterization Guidance

Parameterization is appropriate for the fail-closed matrix where each case is mechanically:

1. Build a standard STS SigV4 request.
2. Mutate one malformed header/query/auth field.
3. Run the real firewall auth path.
4. Assert the local 502 response and expected error fragment.

Each parameterized case should have a clear pytest `id` so failures remain easy to diagnose.

Do not aggressively parameterize every success-path test. These cases should stay readable as standalone behavior tests because they carry important signature and URL semantics:

- Reference AWS signature expectations.
- Encoded path handling.
- Normalized host handling.
- Trusted original URL handling.
- Resolved query preservation.
- Literal plus preservation.

## What Not To Do

Do not move expected final signatures into helpers. Expected signatures should remain in the tests so regressions are visible at the assertion site.

Do not mock away the `real_flow(...)` / `auth.handle_firewall_request(...)` path. The value of these tests is that they exercise the firewall auth integration path with realistic mitmproxy flows.

Do not create one catch-all helper that knows about signer internals, firewall auth config, cache state, fake auth endpoints, and flow assertions. That would reduce duplication but make the tests harder to review.

Do not use module-level mutable dict/list fixtures. Builders should return fresh objects on each call to avoid cross-test mutation.

## Risks

The main risk is reducing test readability by hiding the specific malformed input inside overly generic helpers. Keep helper names and parameters concrete, and keep each test's behavioral difference visible at the call site.

Another risk is accidentally dropping test coverage while reshaping the large file. Use pytest collection before and after the refactor to compare collected test cases.

## Validation

Before and after the refactor, compare collection for the affected tests:

- `cd crates/runner/mitm-addon && uv run pytest --collect-only tests/test_firewall_aws_sigv4_auth.py tests/test_aws_sigv4.py tests/test_firewall_auth.py tests/test_firewall_rewrite_forwarding.py`

Run focused quality checks:

- `cd crates/runner/mitm-addon && uv run ruff check tests/test_firewall_aws_sigv4_auth.py tests/test_aws_sigv4.py tests/test_firewall_auth.py tests/test_firewall_rewrite_forwarding.py tests/aws_sigv4_helpers.py tests/firewall_aws_sigv4_helpers.py`
- `cd crates/runner/mitm-addon && uv run basedpyright tests/aws_sigv4_helpers.py tests/firewall_aws_sigv4_helpers.py tests/test_firewall_aws_sigv4_auth.py tests/test_aws_sigv4.py tests/test_firewall_auth.py tests/test_firewall_rewrite_forwarding.py`
- `cd crates/runner/mitm-addon && uv run pytest tests/test_firewall_aws_sigv4_auth.py tests/test_aws_sigv4.py tests/test_firewall_auth.py tests/test_firewall_rewrite_forwarding.py`

---
> Compacted on 2026-07-09 from 0 comments plus conversation context.

### PR DESCRIPTION
**Pull request by seven332**
## Summary

Closes #20785.

This is a test-only refactor for the mitm-addon AWS SigV4 coverage. It adds two focused helper layers:

- `tests/aws_sigv4_helpers.py` for pure SigV4 fixture vocabulary such as credential scopes, placeholder auth headers, presigned query paths, and test credentials.
- `tests/firewall_aws_sigv4_helpers.py` for firewall-auth integration setup while keeping the real `auth.handle_firewall_request(...)` path.

The large firewall AWS SigV4 test file now uses the shared helpers, while expected final signatures and malformed inputs remain visible at the test sites. Neighboring SigV4 tests were updated only for direct fixture reuse. No production code changed.

## Validation

- `pytest --collect-only -q tests/test_firewall_aws_sigv4_auth.py tests/test_aws_sigv4.py tests/test_firewall_auth.py tests/test_firewall_rewrite_forwarding.py` -> 223 tests collected
- `ruff check tests/aws_sigv4_helpers.py tests/firewall_aws_sigv4_helpers.py tests/test_firewall_aws_sigv4_auth.py tests/test_aws_sigv4.py tests/test_firewall_auth.py tests/test_firewall_rewrite_forwarding.py`
- `basedpyright tests/aws_sigv4_helpers.py tests/firewall_aws_sigv4_helpers.py tests/test_firewall_aws_sigv4_auth.py tests/test_aws_sigv4.py tests/test_firewall_auth.py tests/test_firewall_rewrite_forwarding.py`
- `pytest tests/test_firewall_aws_sigv4_auth.py tests/test_aws_sigv4.py tests/test_firewall_auth.py tests/test_firewall_rewrite_forwarding.py` -> 223 passed

### CODE REVIEW

**Comment by seven332** [comment]
## Code Review: PR #20815

### Summary

Reviewed the AWS SigV4 mitm-addon test fixture refactor. No blocking issues found. The PR centralizes SigV4 request/auth/cache fixtures, keeps the tests on real `HTTPFlow` + real `auth.handle_firewall_request` paths, and adds useful cache identity and malformed SigV4 edge-case coverage.

### Key Findings

#### Critical Issues (P0)

- None.

#### High Priority Issues (P1)

- None.

### Testing Review

#### Coverage

- test refactor: shared AWS SigV4 fixture construction -> covered by the existing and updated `test_firewall_aws_sigv4_auth.py`, `test_aws_sigv4.py`, `test_firewall_auth.py`, and `test_firewall_rewrite_forwarding.py`.
- cache identity behavior -> covered by cache hit/miss cases for auth query, allow context, and billable context.
- malformed/hostile SigV4 source inputs -> covered by fail-closed cases for real source access keys, duplicate credentials, duplicate signed headers, malformed dates, malformed request targets, and content hash ambiguity.

#### Convention Compliance

- No P0/P1 testing convention violations found.
- Uses real mitmproxy `HTTPFlow` fixtures rather than hand-built flow mocks.
- Mocks remain at external or boundary seams (`FakeAuthEndpoint`, auth service fetch/forwarding boundaries in pre-existing test files).
- No fake timers, sleeps, skipped tests, direct fetch mocks, `Any`, lint/type suppressions, or internal `vi.mock()` patterns were introduced.

#### Testing Verdict: Adequate

### Bad Smell Analysis

- Mock violations: 0
- Direct fetch mocking: 0
- Internal code mocking: 0
- TypeScript/Python `Any`: 0
- Lint/type suppressions: 0
- Dynamic imports: 0
- Defensive try/catch blocks: 0
- Artificial sleeps/delays: 0
- Deployment compatibility concerns: 0

### Verification

- `git diff --check origin/main...HEAD`
- `ruff check tests/aws_sigv4_helpers.py tests/firewall_aws_sigv4_helpers.py tests/test_aws_sigv4.py tests/test_firewall_auth.py tests/test_firewall_aws_sigv4_auth.py tests/test_firewall_rewrite_forwarding.py`
- `basedpyright tests/aws_sigv4_helpers.py tests/firewall_aws_sigv4_helpers.py tests/test_aws_sigv4.py tests/test_firewall_auth.py tests/test_firewall_aws_sigv4_auth.py tests/test_firewall_rewrite_forwarding.py`
- `pytest --collect-only -q tests/test_firewall_aws_sigv4_auth.py tests/test_aws_sigv4.py tests/test_firewall_auth.py tests/test_firewall_rewrite_forwarding.py`
- `pytest tests/test_firewall_aws_sigv4_auth.py tests/test_aws_sigv4.py tests/test_firewall_auth.py tests/test_firewall_rewrite_forwarding.py -q`

### Recommendations

- None.

### Verdict

LGTM.

---

Full review details: `codereviews/20260710/pr-20815-commit-list.md`
Testing standards: `docs/testing.md`, `docs/testing/mitm-addon-testing.md`

File: crates/runner/mitm-addon/tests/firewall_aws_sigv4_helpers.py [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+    def __call__(
+        self,
+        *,
+        host: str = ...,
+        path: str = ...,
+        method: str = ...,
+        request_body: bytes | None = ...,
+        request_headers: http.Headers | None = ...,
+        with_response: bool = ...,
+    ) -> http.HTTPFlow: ...
```
[the remark this answers is not in the document]
Comment (seven332) [reply crates/runner/mitm-addon/tests/firewall_aws_sigv4_helpers.py]: Addressed in 58d6422a0b. I used `raise NotImplementedError` instead of `pass` so the protocol stubs no longer have ellipsis-only bodies while still satisfying basedpyright return-type checks.
